File: DFS/078.Subsets/078.Subsets.ts
function subsets(nums: number[]): number[][] {
  const ret = [[]];
  const current = [];
  for (let i = 1; i <= nums.length; i++) {
    dfs(0, i);
  }
  return ret;

  function dfs(index: number, total: number) {
    if (current.length === total) {
      ret.push([...current]);
    }

    for (let i = index; i < nums.length; i++) {
      current.push(nums[i]);
      dfs(i + 1, total);
      current.pop();
    }
  }
}

File: DFS/078.Subsets/078.Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        current = []
        def dfs(index, goalLen):
            if len(current) == goalLen:
                result.append(current[:])
                return
            
            for i in range(index, len(nums)):
                current.append(nums[i])
                dfs(i + 1, goalLen)
                current.pop()
            
        for i in range(len(nums) + 1):
            dfs(0, i)
        
        return result
            
